addon/services/indexed-db.js

```diff
@@ -609,26 +610,36 @@ export default class IndexedDbService extends Service {
       // If a multi index is found, use it
       if (multiIndex) {
         let keyPath = get(multiIndex, 'keyPath');
-        let compareValues = array();
-        keyPath.forEach((key) => {
-          compareValues.push(query[key]);
-        });
+        let compareValues = keyPath.map((key) => query[key]);
 
         let keyName = get(multiIndex, 'name');
         return db[type].where(keyName).equals(compareValues);
       }
     }
 
     // Else, filter manually
-    Object.keys(query).forEach((i) => {
-      if (!promise) {
-        promise = db[type].where(i).equals(query[i]);
-      } else {
-        promise = promise.and((item) => get(item, i) === query[i]);
-      }
-    });
-
-    return promise;
+    return Object.entries(query)
+      .reduce(
```

Comment:
Please avoid using reduce, it is hard to read and follow.

Reply:
@mydea It's your project so I'm happy to change it :)

But! 

I don't necessarily agree. Reduce (or collect as its often called) is self-documenting. When I see reduce I know that you are accumulating a collection into something else, so without ever reading the code I have some context to start with. Whereas when I see a for loop iterating over some collection, it could be doing any number of things, with all sorts of side effects.

Another benefit of using reduction is that you have a pure function that accumulates the query, in general, I find pure functions are easier to reason about as they only have inputs and outputs, and no side effects, which makes writing tests for them a bit less onerous. 

I notice you do a lot of ```forEach``` loops where you append to a list, in those cases why not use map/filter? When I see map, I know that I'm transforming a single list into another list of the same size, but potentially different type.

Again, it's your project, and I know people get tripped up on reduce, and some higher-order functions in general, just some food for thought :)

Reply:
Yes, there are for sure some places where refactoring to `map()` makes sense :)
Regarding `reduce()`, I try to only use it for very short things, e.g. when summing a number. For more complex things (like this), I think it obscures a bit what is going on.

I think it is also OK to leave it with reduce, and add some comments about what is going on and why. Especially as the Dexie/IndexedDb syntax can be a bit tricky, so it is not that obvious why certain things happen :)